# Guided Project: Exploring Ebay Car Sales Data
The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import datetime as dt

In [2]:
autos = pd.read_csv('./data/autos.csv', encoding='Latin-1')

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

## テーブル

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

## データクリーニング

In [5]:
autos = autos.rename(columns = {'yearOfRegistration': 'registration_year', 
                                'monthOfRegistration': 'registration_month',
                                'notRepairedDamage': 'unrepaired_damage',
                                'dateCreated': 'ad_created',
                                'dateCrawled': 'date_crawled',
                                'offerType': 'offer_type',
                                'vehicleType': 'vehicle_type',
                                'powerPS': 'power_ps',
                                'fuelType': 'fuel_type',
                                'nrOfPictures': 'nr_pictures',
                                'postalCode': 'postal_code',
                                'lastSeen': 'last_seen'}
                    )

In [6]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [7]:
autos['price'] = autos['price'].str.replace(',', '')
autos['price'] = autos['price'].str.replace('$','').astype(int)
autos['odometer_km'] = autos['odometer'].str.replace(',', '')
autos['odometer_km'] = autos['odometer_km'].str.replace('km', '').astype(int)

In [8]:
autos.describe()

price  registration_year      power_ps  registration_month  \
count  5.000000e+04       50000.000000  50000.000000        50000.000000   
mean   9.840044e+03        2005.073280    116.355920            5.723360   
std    4.811044e+05         105.712813    209.216627            3.711984   
min    0.000000e+00        1000.000000      0.000000            0.000000   
25%    1.100000e+03        1999.000000     70.000000            3.000000   
50%    2.950000e+03        2003.000000    105.000000            6.000000   
75%    7.200000e+03        2008.000000    150.000000            9.000000   
max    1.000000e+08        9999.000000  17700.000000           12.000000   

       nr_pictures   postal_code    odometer_km  
count      50000.0  50000.000000   50000.000000  
mean           0.0  50813.627300  125732.700000  
std            0.0  25779.747957   40042.211706  
min            0.0   1067.000000    5000.000000  
25%            0.0  30451.000000  125000.000000  
50%            0.0  49577.000000  150000.000000  
75%            0.0  71540.000000  150000.000000  
max            0.0  99998.000000  150000.000000

- nr_picturesは値が全て０
- unrepaired_damageは(nein: 35232)と(ja: 4939)
- sellerは(privat: 49999)と(gewerblich: 1)
- offer_typeは(Angebot: 49999)と(Gesuch: 1)
- abtestは(test       25756)(control    24244)


外れ値と判断できるレコードは削除する
- priceの０ドルのレコードは削除する
- priceが5000000ドル以上のレコードは削除する

In [22]:
autos['price'] = autos.loc[autos['price'].between(0, 5000000),  'price']
autos = autos.dropna()

datetimeのデータをdateに変換する

In [10]:
autos['date_crawled'] = pd.to_datetime(autos['date_crawled']).dt.date
autos['ad_created'] = pd.to_datetime(autos['ad_created']).dt.date
autos['last_seen'] = pd.to_datetime(autos['last_seen']).dt.date

日付データの分布
- date_crawled
- ad_created
- last_seen

In [11]:
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.00142
2016-04-06    0.00318
2016-04-05    0.01310
2016-04-04    0.03652
2016-04-03    0.03868
2016-04-02    0.03540
2016-04-01    0.03380
2016-03-31    0.03192
2016-03-30    0.03362
2016-03-29    0.03418
2016-03-28    0.03484
2016-03-27    0.03104
2016-03-26    0.03248
2016-03-25    0.03174
2016-03-24    0.02910
2016-03-23    0.03238
2016-03-22    0.03294
2016-03-21    0.03752
2016-03-20    0.03782
2016-03-19    0.03490
2016-03-18    0.01306
2016-03-17    0.03152
2016-03-16    0.02950
2016-03-15    0.03398
2016-03-14    0.03662
2016-03-13    0.01556
2016-03-12    0.03678
2016-03-11    0.03248
2016-03-10    0.03212
2016-03-09    0.03322
2016-03-08    0.03330
2016-03-07    0.03596
2016-03-06    0.01394
2016-03-05    0.02538
Name: date_crawled, dtype: float64

In [12]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.00128
2016-04-06    0.00326
2016-04-05    0.01184
2016-04-04    0.03688
2016-04-03    0.03892
               ...   
2015-12-05    0.00002
2015-11-10    0.00002
2015-09-09    0.00002
2015-08-10    0.00002
2015-06-11    0.00002
Name: ad_created, Length: 76, dtype: float64

In [13]:
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.13092
2016-04-06    0.22100
2016-04-05    0.12428
2016-04-04    0.02462
2016-04-03    0.02536
2016-04-02    0.02490
2016-04-01    0.02310
2016-03-31    0.02384
2016-03-30    0.02484
2016-03-29    0.02234
2016-03-28    0.02086
2016-03-27    0.01602
2016-03-26    0.01696
2016-03-25    0.01920
2016-03-24    0.01956
2016-03-23    0.01858
2016-03-22    0.02158
2016-03-21    0.02074
2016-03-20    0.02070
2016-03-19    0.01574
2016-03-18    0.00742
2016-03-17    0.02792
2016-03-16    0.01644
2016-03-15    0.01588
2016-03-14    0.01280
2016-03-13    0.00898
2016-03-12    0.02382
2016-03-11    0.01252
2016-03-10    0.01076
2016-03-09    0.00986
2016-03-08    0.00760
2016-03-07    0.00536
2016-03-06    0.00442
2016-03-05    0.00108
Name: last_seen, dtype: float64

registration_yearの分布

In [23]:
autos['registration_year'].describe()

count    35045.000000
mean      2003.351947
std          6.455596
min       1931.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2018.000000
Name: registration_year, dtype: float64